core/commands/dag/dag.go

```diff
@@ -241,3 +261,248 @@ var DagResolveCmd = &cmds.Command{
 	},
 	Type: ResolveOutput{},
 }
+
+type importResult struct {
+	roots map[cid.Cid]bool
+	err   error
+}
+
+var DagImportCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Import the contents of .car files",
+		ShortDescription: `
+'ipfs dag import' parses .car files and adds all objects to the blockstore.
+Currently only car files with version 1 are supported, without particular
+hard limits on the contents (so called car-determinism is not enforced).
+
+By default, after all supplied (one or more) car files have been processed,
+an attempt is made to pin recursively each individual root specified in any car
+headers. This is done at the end of the import, allowing pinning to take place
+over the contents of all cars (a dag scattered over multiple files) plus
+anything present in the blockstore.
+
+During pinning no network requests are made, with no option to change that.
+`,
+	},
+	Arguments: []cmds.Argument{
+		cmds.FileArg("path", true, true, "The path of a .car file.").EnableStdin(),
+	},
+	Options: []cmds.Option{
+		cmds.BoolOption(silentOptionName, "No output."),
+		cmds.BoolOption(pinRootsOptionName, "Pin optional roots listed in the .car headers after importing.").WithDefault(true),
+	},
+	Type: CarImportOutput{},
+	Run: func(req *cmds.Request, res cmds.ResponseEmitter, env cmds.Environment) error {
+
+		node, err := cmdenv.GetNode(env)
+		if err != nil {
+			return err
+		}
+
+		api, err := cmdenv.GetApi(env, req)
+		if err != nil {
+			return err
+		}
+
+		// on import ensure we do not reach out to the network for any reason
+		// if a pin based on what is imported + what is in the lockstore
+		// isn't possible: tough luck
+		api, _ = api.WithOptions(options.Api.Offline(true))
+
+		// grab a gc lock so that regardless of the size of the streamed-in cars
+		// nothing will disappear on us before we had a chance to pin the roots
+		// at the very end
+		// This is especially important for use cases like dagger:
+		//    ipfs dag import $( ... | ipfs-dagger --stdout=carfifos )
+		//
+		unlocker := node.Blockstore.PinLock()
+		defer func() { unlocker.Unlock() }()
+
+		doPinRoots, _ := req.Options[pinRootsOptionName].(bool)
+
+		retCh := make(chan importResult, 1)
+		go importWorker(req, res, api, retCh)
+
+		done := <-retCh
+		if done.err != nil {
+			return done.err
+		}
+
+		// It is not guaranteed that a root in a header is actually present in the same ( or any )
+		// .car file. This is the case in version 1, and ideally in further versions too
+		// Accumulate any root CID seen in a header, and supplement its actual node if/when encountered
+		// We will attempt a pin *only* at the end in case all car files were well formed
+		//
+		// The boolean value indicates whether we have encountered the root within the car file's
+		//var roots map[cid.Cid]bool
```

Comment:
is this used as a hint, or intentional?
if its a hint to help the reader understand the type, it might be better to have it in code so it stays in sync
```
var roots map[cid.Cid]bool
roots = done.roots
```

Reply:
Fixed